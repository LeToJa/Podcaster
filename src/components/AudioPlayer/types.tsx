export interface AudioPlayerTypes {
    url: string
    type: string
}