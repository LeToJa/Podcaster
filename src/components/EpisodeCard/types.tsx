export interface EpisodeCardTypes {
    podcastId: string
    episodeId: number
    title: string
    date: string
    duration: string
}