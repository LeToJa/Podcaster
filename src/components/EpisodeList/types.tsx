import { PodcastEpisode } from '../../layout/PodcastLayout/types'

export interface EpisodeListTypes {
    podcastId: string
    episodes: PodcastEpisode[]
}