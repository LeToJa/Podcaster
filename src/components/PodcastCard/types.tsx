export interface PocastCardTypes {
    id: string
    name: string
}