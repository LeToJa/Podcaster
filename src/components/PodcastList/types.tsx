import { PocastCardTypes } from '../PodcastCard/types'

export interface PodcastListTypes {
    podcasts: PocastCardTypes[]
}