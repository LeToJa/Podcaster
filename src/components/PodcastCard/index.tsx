import { Link } from 'react-router-dom'

import { PocastCardTypes } from './types'

const PodcastCard = ({ id, title, author, artwork }: PocastCardTypes) => <>
    <Link to={`/podcast/${id}`} className='block shadow-xl px-2 py-4'>
        <img className='w-2/5 mx-auto rounded-full' src={artwork} alt={title} />
        <header className='px-6 py-4'>
            <h4 className='font-bold text-xl mb-2 text-center whitespace-nowrap text-ellipsis overflow-hidden'>{title}</h4>
            <h6 className='text-gray-700 text-base text-center whitespace-nowrap text-ellipsis overflow-hidden'>{author}</h6>
        </header>
    </Link>
</>

export default PodcastCard