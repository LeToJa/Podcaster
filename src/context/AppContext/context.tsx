import { createContext, useState } from 'react'

import { ReactElementWithChildrenTypes } from '../../types'
import { AppContextTypes } from './types'

export const AppContext = createContext<AppContextTypes | undefined>(undefined)

export const AppProvider = ({ children }: ReactElementWithChildrenTypes) => {
    const [loading, setLoading] = useState<boolean>(true)

    const toggleLoading = (value: boolean) => setLoading(value)

    return <>
        <AppContext.Provider value={{ loading, toggleLoading }}>
            {children}
        </AppContext.Provider>
    </>
}