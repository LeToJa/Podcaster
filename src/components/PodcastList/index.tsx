import { useState } from 'react'

import PodcastCard from '../../components/PodcastCard'

import { PodcastListTypes } from './types'

const PodcastList = ({ podcasts }: PodcastListTypes) => {

    const [filter, setFilter] = useState<string>('')


    const filteredPodcasts = podcasts?.filter(podcasts =>
        podcasts.title.toLowerCase().includes(filter.toLowerCase()) ||
        podcasts.author.toLowerCase().includes(filter.toLowerCase())
    )

    return <>
        <form className='flex justify-end items-center gap-4 mb-6' data-testid='app-form'>
            <label className='px-1 bg-sky-600 text-white font-bold rounded-lg' data-testid='form-label'>
                {filteredPodcasts.length}
            </label>
            <input
                type='text'
                placeholder='Search podcasts'
                name='filter-podcasts'
                id='filter-podcasts'
                className='bg-gray-100 px-4 py-3 rounded shadow-md outline-sky-600'
                value={filter}
                onChange={(e) => setFilter(e.target.value)}
                data-testid='form-input'
            />
        </form>
        <section className='grid grid-cols-5 gap-4'>
            {
                filteredPodcasts.length ?
                    filteredPodcasts?.map((podcast, key) => <PodcastCard key={key} {...podcast} />) :
                    <section className='col-span-5 font-bold text-xl text-gray-700 text-center' data-testid='app-message'>No podcasts :(</section>
            }
        </section>
    </>
}

export default PodcastList