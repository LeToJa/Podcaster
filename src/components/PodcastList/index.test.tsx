import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import { describe, it, expect, vi } from 'vitest'
import { BrowserRouter } from 'react-router-dom'

import { AppContext } from '../../context/AppContext/context'
import { AppContextProps } from '../../context/AppContext/types'
import { ReactElementWithChildrenTypes } from '../../types'
import { PocastCardTypes } from '../PodcastCard/types'

import PodcastList from '.'

const mockAppContext = {
    loading: false,
    toggleLoading: vi.fn()
}

const MockAppProvider = ({ loading, toggleLoading, children }: AppContextProps & ReactElementWithChildrenTypes) => {
    return (
        <AppContext.Provider value={{ loading, toggleLoading }}>
            {children}
        </AppContext.Provider>
    )
}

const mockPodcasts: PocastCardTypes[] = [
    {
        id: '1',
        title: 'Podcast One',
        author: 'Author One',
        artwork: 'https://via.placeholder.com/150'
    },
    {
        id: '2',
        title: 'Podcast Two',
        author: 'Author Two',
        artwork: 'https://via.placeholder.com/150'
    },
]

describe('Home', () => {
    it('renders correctly and matches snapshot', async () => {
        const { container, getByText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <PodcastList podcasts={mockPodcasts} />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText(/Podcast One/i)).toBeTruthy()
        })

        expect(container).toMatchSnapshot()
    })

    it('displays podcasts', async () => {
        const { getByText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <PodcastList podcasts={mockPodcasts} />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText('Podcast One')).toBeTruthy()
            expect(getByText('Podcast Two')).toBeTruthy()
        })
    })

    it('filters podcasts based on user input', async () => {
        const { getByText, queryByText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <PodcastList podcasts={mockPodcasts} />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText('Podcast One')).toBeTruthy()
            expect(getByText('Podcast Two')).toBeTruthy()
        })

        fireEvent.change(screen.getByPlaceholderText('Search podcasts'), {
            target: { value: 'One' },
        })

        expect(getByText('Podcast One')).toBeTruthy()
        expect(queryByText('Podcast Two')).toBeFalsy()
    })

    it('displays "No podcasts :(" when no podcasts match the filter', async () => {
        const { getByText, getByPlaceholderText } = render(
            <BrowserRouter>
                <MockAppProvider {...mockAppContext}>
                    <PodcastList podcasts={mockPodcasts} />
                </MockAppProvider>
            </BrowserRouter>
        )

        await waitFor(() => {
            expect(getByText('Podcast One')).toBeTruthy()
            expect(getByText('Podcast Two')).toBeTruthy()
        })

        fireEvent.change(getByPlaceholderText('Search podcasts'), {
            target: { value: 'Nonexistent Podcast' },
        })

        expect(getByText('No podcasts :(')).toBeTruthy()
    })
})
