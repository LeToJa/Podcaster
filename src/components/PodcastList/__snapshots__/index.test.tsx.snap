// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Home > renders correctly and matches snapshot 1`] = `
<div>
  <form
    class="flex justify-end items-center gap-4 mb-6"
    data-testid="app-form"
  >
    <label
      class="px-1 bg-sky-600 text-white font-bold rounded-lg"
      data-testid="form-label"
    >
      2
    </label>
    <input
      class="bg-gray-100 px-4 py-3 rounded shadow-md outline-sky-600"
      data-testid="form-input"
      id="filter-podcasts"
      name="filter-podcasts"
      placeholder="Search podcasts"
      type="text"
      value=""
    />
  </form>
  <section
    class="grid grid-cols-5 gap-4"
  >
    <a
      class="block bg-gray-50 border-2 border-gray-200 shadow-xl px-2 py-4"
      data-testid="podcast-card"
      href="/podcast/1"
    >
      <img
        alt="Podcast One"
        class="w-2/5 mx-auto rounded-full"
        src="https://via.placeholder.com/150"
      />
      <header
        class="px-6 py-4"
      >
        <h4
          class="font-bold text-xl mb-2 text-center whitespace-nowrap text-ellipsis overflow-hidden"
        >
          Podcast One
        </h4>
        <h6
          class="text-gray-700 text-base text-center whitespace-nowrap text-ellipsis overflow-hidden"
        >
          Author One
        </h6>
      </header>
    </a>
    <a
      class="block bg-gray-50 border-2 border-gray-200 shadow-xl px-2 py-4"
      data-testid="podcast-card"
      href="/podcast/2"
    >
      <img
        alt="Podcast Two"
        class="w-2/5 mx-auto rounded-full"
        src="https://via.placeholder.com/150"
      />
      <header
        class="px-6 py-4"
      >
        <h4
          class="font-bold text-xl mb-2 text-center whitespace-nowrap text-ellipsis overflow-hidden"
        >
          Podcast Two
        </h4>
        <h6
          class="text-gray-700 text-base text-center whitespace-nowrap text-ellipsis overflow-hidden"
        >
          Author Two
        </h6>
      </header>
    </a>
  </section>
</div>
`;
